### PR TITLE
Ext2 write + growing inode support

### PR DESCRIPTION
**Pull request by Supercip971**

### CODE REVIEW

File: sources/libs/embed/posix/file.c [hunk 1]
```diff
@@ -9,11 +9,27 @@
...
+    switch(flags)
+    {
+        case FILE_OPEN_READ_ONLY:
+            return O_RDONLY;
+        case FILE_OPEN_WRITE_ONLY:
+            return O_WRONLY;
+        case FILE_OPEN_READ_WRITE:
+            return O_RDWR;
+        default:
+            panic$("unkown embed flag: {}", flags);
```
Comment (sleepy-monax) [inline sources/libs/embed/posix/file.c:24]: Unkown open flag: {}

File: sources/libs/fs/ext2/ext2.c [hunk 2]
```diff
@@ -2,12 +2,47 @@
...
+static Ext2CachedInode *get_cached_inode_or_create(Ext2Fs *self, Ext2FsInode inode)
+{
+    Ext2CachedInode *cached = get_cached_inode(self, inode.id);
+    if (cached == nullptr)
+    {
+        return add_cached_inode(self, inode);
+    }
+
+    return cached;
+}
```
Comment (sleepy-monax) [inline sources/libs/fs/ext2/ext2.c:38]: The inode cache should live it's own translation unit

File: sources/libs/fs/ext2/ext2.c [hunk 3]
```diff
@@ -43,120 +100,417 @@ static size_t get_entry_in_indirect_block(size_t entry_per_block, int depth)
...
-    if (super_block->read_only_feature & EXT2_FEAT_64BIT_FILE_SIZE)
+    if (self->super_block->read_only_feature & EXT2_FEAT_64BIT_FILE_SIZE)
     {
         len |= (uint64_t)inode->block.size_upper << 32;
     }
 
     return len;
 }
 
+FsResult ext2_inode_rw(Ext2Fs *self, Ext2FsInode *inode, Ext2FsInodeId id, bool write)
```
Comment (sleepy-monax) [inline sources/libs/fs/ext2/ext2.c:114]: I don't like function behaving completely differently based on a boolean, most of the time this mean the fonctions need to be splitted

File: sources/libs/fs/ext2/ext2.c [hunk 4]
```diff
@@ -43,120 +100,417 @@ static size_t get_entry_in_indirect_block(size_t entry_per_block, int depth)
...
+    *result = bits_find_free(&bit, 0, 1, false).base;
+
+    bits_set(&bit, *result, true);
+
+    ext2_release_blocks(self, bitmap);
+
+    return FS_SUCCESS;
+}
+
+static FsResult ext2_alloc_block(Ext2Fs *self, uint32_t preferred_group_id, uint32_t *result)
```
Comment (sleepy-monax) [inline sources/libs/fs/ext2/ext2.c:338]: block allocation should live in it's own TU

File: sources/libs/fs/fs.h [hunk 5]
```diff
@@ -8,4 +8,5 @@ typedef enum
     FS_INVALID_INODE,
     FS_INVALID_PATH,
     FS_INVALID_FS,
+    FS_NO_FREE_BLOCK
 } FsResult;
```
Comment (sleepy-monax) [inline sources/libs/fs/fs.h:12]: The Fs subsystem should use IoResult, making interop between our IO and FS subsystem easier.

**Review by sleepy-monax — CHANGES_REQUESTED** [review]

File: sources/libs/fs/ext2/ext2.c (hunk 3 above)
Comment (Supercip971) [reply sources/libs/fs/ext2/ext2.c:114]: It's to avoid repetitive code, and this function is splitted in 2: ext2_write_inode & ext2_inode

Comment (sleepy-monax) [reply sources/libs/fs/ext2/ext2.c:114]: Well then just isolate the common code in another function

Comment (Supercip971) [reply sources/libs/fs/ext2/ext2.c:114]: I don't see the point of doing it if in the end we only use it in 2 functions.  (it should be static tho)

**Pull request merged**